File: leetcode/path-sum-ii.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if not root: return []

        st = []
        res = []
        root.parent = None
        st.append((root, sum, 0))

        while st:
            (r, v, d) = st.pop()
            if not r:
                continue

            if r.left is None and r.right is None:
                if r.val == v:
                    t = r
                    path = []
                    while t:
                        path.append(t.val)
                        t = t.parent
                    res.append(path[::-1])
                continue

            if d == 0:
                st.append((r, v, d + 1))
                ln = r.left
                if ln: ln.parent = r
                st.append((ln, v - r.val, 0))

            elif d == 1:
                rn = r.right
                if rn: rn.parent = r
                st.append((rn, v - r.val, 0))

        return res

File: misc/leetcode/cpp/path-sum-ii.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include <cstdio>
#include <vector>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left, *right;
};

class Solution {
   public:
    void trace(TreeNode* root, int sum, vector<int>& stage,
               vector<vector<int>>& paths) {
        if (root == NULL) return;
        if (root->left == NULL && root->right == NULL && root->val == sum) {
            stage.push_back(root->val);
            paths.push_back(stage);
            stage.pop_back();
            return;
        }
        stage.push_back(root->val);
        trace(root->left, sum - root->val, stage, paths);
        trace(root->right, sum - root->val, stage, paths);
        stage.pop_back();
    }
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> paths;
        vector<int> stage;
        trace(root, sum, stage, paths);
        return paths;
    }
};
